working/0110-incremental-sound-nnbd/feature-specification.md

```diff
@@ -0,0 +1,374 @@
+feature. When weak null checking is enabled, all errors specified this proposal
+(that is, all errors that arise only out of the new features of this proposal)
+shall be treated as warnings.
+
+Strong null checking is enabled by running the compilation or execution
+environment with the appropriate flags.  When strong null checking is enabled,
+errors specified in this proposal shall be treated as errors.
+
+### Legacy libraries
+
+Static checking for a library which has not opted into this feature (a *legacy*
+library) is done using the semantics as of the last version of the language
+before this feature ships (or the last version to which it has opted in, if that
+is different).  All opted-in libraries upstream from the legacy library are
+viewed by the legacy library with nullability related features erased from their
+APIs.  In particular:
+  - All types of the form `T?` in the opted-in API are treated as `T`.
+  - All required named parameters are treated as optional named parameters.
+  - The type `Never` is treated as the type `Null`
+
+In a legacy library, none of the new syntax introduced by this proposal is
+available, and it is a static error if it is used.
+
+### Importing legacy libraries from opted-in libraries
+
+The type system is extended with a notion of a legacy type operator.  For every
+type `T`, there is an additional type `T*` which is the legacy version of the
+type.  There is no surface syntax for legacy types, and implementations should
+display the legacy type `T*` in the same way that they would display the type
+`T`, except in so far as it is useful to communicate to programmers for the
+purposes of error messages that the type originates in unmigrated code.
+
+When static checking is done in a migrated library, types which are imported
+from unmigrated libraries are seen as legacy types.  However, for the purposes
+of type inference in migrated libraries, types imported from unmigrated
+libraries shall be treated as non-nullable.  As a result, legacy types will
+never appear as type annotations in migrated libraries, nor will they appear in
+reified positions.
+
+### Type reification
+
+All types reified in legacy libraries are reified as legacy types.  Runtime
+subtyping checks treat them according to the subtyping rules specified
+separately.
+
+### Runtime checks and weak checking
+
+When weak checking is enabled, runtime type tests (including explicit and
+implicit casts) shall succeed with a warning whenever the runtime type test
+would have succeeded if all `?` types were ignored, `Never` were treated as
+`Null`, and `required` named parameters were treated as optional.
+
+### Exports
+
+If an unmigrated library re-exports a migrated library, the re-exported symbols
+retain their migrated status (that is, downstream migrated libraries will see
+their migrated types).
+
+It is an error for a migrated library to re-export symbols from an unmigrated
+library.
```

Comment:
This can hold back migration. 
Also, if an unmigrated library can export migrated "symbols", this might not even export any unmigrated types.

Maybe:
It is an error for a migrated library to re-export declarations with a static type containing unmigrated types.

Reply:
Do you really think this is likely?  Migrating before your upstream packages migrate is already the less common use case. Exporting symbols is a fringe use case.  Exporting symbols that were exported from yet another library seems like a fringe on a fringe case.  It seems much simpler to me to say that we just don't support it.

Reply:
Why is it that we won't support it? That is, what is the problem we are trying to solve by disallowing the export?

Is it purely methodological problems: Library A is migrated, but exports declarations from unmigrated library B. Library C uses library A and wants to migrate, so they might need to migrate twice, once for A and once more when B is migrated. We don't want that, so we disallow A from exporting declarations from B until B has been migrated.

If so, I'm not sure how that differs from an unmigrated library exporting declarations of a migrated library. If library A exports declarations from library B and library B migrates first, then the user in library C, which imports library A, may need to migrate twice as well, once for the B changes, and once more for the A changes.

They can choose to not migrate library C until both A and B have been migrated, but that applies to both cases.

Or is there a technical problem? I can't see one because you can get into the exact same state by importing library A and B separately, rather than relying on the export of B by A, and we need to be able to handle that.

Reply:
It seems to me to add a lot of complication if migrated libraries can have subset of their APIs which are unmigrated.